### PR TITLE
Check that pointers to SuiteSparse objects haven't been zeroed before calling SuiteSparse routines. 

### LINKED ISSUE #284: Segmentation fault while fetching for product of sparse qrfact and dense matrix
I was comparing LU- and QR-Factorization and wanted to implement parallelism. I noticed that fetching the QR-factorized matrix `fa` at the same time as the identity matrix `b` results in a segmentation fault, while the same works fine with `lufact`. Additionally, serial code works, too.

```
julia> addprocs(2)
2-element Array{Int64,1}:
 2
 3

julia> @everywhere n = 5

julia> a = @spawn sprand(n,n,0.99)
RemoteRef{Channel{Any}}(2,1,5)

julia> b = @spawn eye(n)
RemoteRef{Channel{Any}}(3,1,6)

julia> fa = @spawn qrfact(fetch(a))
RemoteRef{Channel{Any}}(2,1,7)

julia> c = @spawn fetch(fa) \ fetch(b)
RemoteRef{Channel{Any}}(3,1,8)
```

Error:

```
signal (11): Segmentation fault
size at abstractarray.jl:53
Worker 3 terminated.
ERROR (unhandled task failure): EOFError: read end of file
 in read at stream.jl:911
 in message_handler_loop at multi.jl:863
 in process_tcp_streams at multi.jl:852
 in anonymous at task.jl:63
```

---

```
julia> versioninfo()
Julia Version 0.4.1
Commit cbe1bee (2015-11-08 10:33 UTC)
Platform Info:
  System: Linux (x86_64-unknown-linux-gnu)
  CPU: Intel(R) Core(TM) i5 CPU         750  @ 2.67GHz
  WORD_SIZE: 64
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Nehalem)
  LAPACK: libopenblas
  LIBM: libm
  LLVM: libLLVM-3.3
```

### LINKED ISSUE #14134: Segmentation fault while fetching for product of sparse qrfact and dense matrix
I was comparing LU- and QR-Factorization and wanted to implement parallelism. I noticed that fetching the QR-factorized matrix `fa` at the same time as the identity matrix `b` results in a segmentation fault, while the same works fine with `lufact`. Additionally, serial code works, too.

```
julia> addprocs(2)
2-element Array{Int64,1}:
 2
 3

julia> @everywhere n = 5

julia> a = @spawn sprand(n,n,0.99)
RemoteRef{Channel{Any}}(2,1,5)

julia> b = @spawn eye(n)
RemoteRef{Channel{Any}}(3,1,6)

julia> fa = @spawn qrfact(fetch(a))
RemoteRef{Channel{Any}}(2,1,7)

julia> c = @spawn fetch(fa) \ fetch(b)
RemoteRef{Channel{Any}}(3,1,8)
```

Error:

```
signal (11): Segmentation fault
size at abstractarray.jl:53
Worker 3 terminated.
ERROR (unhandled task failure): EOFError: read end of file
 in read at stream.jl:911
 in message_handler_loop at multi.jl:863
 in process_tcp_streams at multi.jl:852
 in anonymous at task.jl:63
```
---

```
julia> versioninfo()
Julia Version 0.4.1
Commit cbe1bee (2015-11-08 10:33 UTC)
Platform Info:
  System: Linux (x86_64-unknown-linux-gnu)
  CPU: Intel(R) Core(TM) i5 CPU         750  @ 2.67GHz
  WORD_SIZE: 64
  BLAS: libopenblas (DYNAMIC_ARCH NO_AFFINITY Nehalem)
  LAPACK: libopenblas
  LIBM: libm
  LLVM: libLLVM-3.3
```

### PR DESCRIPTION
**Pull request by andreasnoack**
Fixes JuliaLang/LinearAlgebra.jl#284

### CODE REVIEW

**Comment by tkelman** [comment]
Could this use a more descriptive name than just `check` ? At least `checkptr` or `checknull` or `checknullptr` or something to that effect?

**Comment by amitmurthy** [comment]
Wouldn't it be better to

- wrap the pointer in a `Nullable` type.
- consequently use `get` on the value everywhere which automatically throws an exception if it is NULL
- serialization of such a `Nullable{Ptr}` will throw an exception as should be the case.
- In case you do want to allow serialization of other non-pointer fields you will need to write type specific `serialize`/`deserialize` methods.

**Comment by amitmurthy** [comment]
Actually defining a `serialize{T<:Ptr}(::SerializationState, ::Nullable{T})` which would serialize any pointer as NULL will prevent any exceptions being thrown while serialization. It will be only thrown only upon access.

**Comment by andreasnoack** [comment]
`Ptr` is already nullable, right? So wouldn't it be more stright forward to define `get(p::Ptr) = p==C_NULL ? throw(NullException()) : p`? and then use `get` instead of `check`?

The problem is that most users will probably have no clue about why they got a `NullException`. They just wanted to solve a sparse system of equations using several processes and depending on the order of execution of the `@spawn`s, they will either get the expected result or a `NullException` exception without further explanations.

**Comment by amitmurthy** [comment]
While your method works, wrapping a `Ptr` in a `Nullable` type will force usage of `get(::Nullable{Ptr})` at all callsites. Since for a `Nullable`, a `Ptr` value of `0` is a valid value (we want `get` to throw an exception and not return 0), serializing an object of type `Nullable{Ptr}` should actually serialize the value `Nullable{Ptr}()`

An alternate solution that also has better error reporting:

Define a new RemotePtr type.
```
type RemotePtr
  p::Ptr
  pid::Int
end
get(rp::RemotePtr) = (myid() == rp.pid ? rp.p : error("Invalid pointer. This pointer is valid only on pid $(rp.p)"))
```

`serialize(::SerializatioState, ::RemotePtr)` can serialize the actual pointer value instead of 0. This will allow objects to be serialized across workers without any issues and an exception is thrown only upon access.

**Comment by andreasnoack** [comment]
@amitmurthy Thanks for the explanation. I've ended up defining `get{T<:SuiteSparseStruct}(Ptr{T})` such that I treat  pointers similar to `Nullable` nullables. It doesn't enforce the use of `get`, but I've wrapped all pointer accesses in `cholmod.jl` and `spqr.jl` in `get` and we can enforce this in future code reviews. It also allows for simple definition of an informative error message.

The UMFPACK wrappers are a bit different: they check for `C_NULL` and recomputes the factorization in that case. They can do that because they have copies of the sparse matrix in the type.

File: base/sparse/cholmod.jl [hunk 1]
```diff
@@ -417,14 +430,14 @@ function norm_dense{Tv<:VTypes}(D::Dense{Tv}, p::Integer)
...
-          D.p, p, common())
+          get(D.p), p, common())
 end
 
 ### cholmod_check.h ###
 function check_dense{T<:VTypes}(A::Dense{T})
     ccall((:cholmod_l_check_dense, :libcholmod), Cint,
           (Ptr{C_Dense{T}}, Ptr{UInt8}),
-          A.p, common())!=0
+          A.p, common()) == 1
```
Comment (tkelman) [inline base/sparse/cholmod.jl]: unrelated?

Comment (andreasnoack) [inline base/sparse/cholmod.jl]: Yes, it shouldn't matter, but I'll revert it.

Comment (tkelman) [inline base/sparse/cholmod.jl]: as a separate commit would be fine, what's the reasoning?

Comment (andreasnoack) [inline base/sparse/cholmod.jl]: The function is only supposed to return zero or one so it shouldn't matter, but I thought a "is true" test was prettier than a "not false".

**Comment by andreasnoack** [comment]
@tkelman I think this is safe to backport. It's a pure bugfix that changes a segfault to an error. This is what is causing https://github.com/JuliaLang/julia/issues/14355

**Pull request merged**